Frameworks/CoreGraphics/CGContext.mm

```diff
@@ -1238,3 +1238,45 @@ void CGContextDrawGlyphRun(CGContextRef ctx, const DWRITE_GLYPH_RUN* glyphRun) {
 void _CGContextSetScaleFactor(CGContextRef ctx, float scale) {
     ctx->Backing()->_CGContextSetScaleFactor(scale);
 }
+
+#pragma region CGContextBeginDrawEndDraw
+
+void _CGContextPushBeginDraw(CGContextRef ctx) {
+    if ((ctx->_beginEndDrawDepth)++ == 0) {
+        ID2D1RenderTarget* imgRenderTarget = ctx->Backing()->DestImage()->Backing()->GetRenderTarget();
+        THROW_HR_IF_NULL(E_UNEXPECTED, imgRenderTarget);
+        imgRenderTarget->BeginDraw();
+    }
+}
+
+void _CGContextPopEndDraw(CGContextRef ctx) {
+    if (--(ctx->_beginEndDrawDepth) == 0) {
+        ID2D1RenderTarget* imgRenderTarget = ctx->Backing()->DestImage()->Backing()->GetRenderTarget();
+        THROW_HR_IF_NULL(E_UNEXPECTED, imgRenderTarget);
+        THROW_IF_FAILED(imgRenderTarget->EndDraw());
+    }
+}
+
+void _CGContextSafeInnerBeginDraw(CGContextRef ctx) {
```

Comment:
Not used in this commit, but are needed for CGD2D. it seemed vaguely more appropriate to add them now than to add them when merging to CGD2D, but i don't feel strongly about this.

As an example, CGD2D has __CGContext::DrawToCommandList - it switches render targets in the middle and switches back at the end. Retargeting is illegal if in the middle of a Begin/EndDraw pair, so we need an alternate mechanism to force an EndDraw so we can retarget, then BeginDraw again. (I made a mistake by putting the second command in the body of this function though, oops) According to @DHowett-MSFT, shadows and transparency layers also require a retarget.

---
In reply to: [96086173](https://github.com/Microsoft/WinObjC/pull/1705#discussion_r96086173) [](ancestors = 96086173)